test/test_selectors.py

```diff
@@ -0,0 +1,684 @@
+
+    def test_select_no_event_types(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        self.assertRaises(ValueError, s.register, rd, 0)
+
+    def test_select_many_events(self):
+        s = self.make_selector()
+        readers = []
+        writers = []
+        for i in range(32):
+            rd, wr = self.make_socketpair()
+            readers.append(rd)
+            writers.append(wr)
+            s.register(rd, selectors.EVENT_READ)
+
+        self.assertEqual(0, len(s.select(0.001)))
+
+        # Write a byte to each end.
+        for wr in writers:
+            wr.send(b'x')
+
+        # Give time to flush the writes.
+        time.sleep(0.01)
+
+        ready = s.select(0.001)
+        self.assertEqual(32, len(ready))
+        for key, events in ready:
+            self.assertEqual(selectors.EVENT_READ, events)
+            self.assertIn(key.fileobj, readers)
+
+        # Now read the byte from each endpoint.
+        for rd in readers:
+            data = rd.recv(1)
+            self.assertEqual(b'x', data)
+
+        self.assertEqual(0, len(s.select(0.001)))
+
+    def test_select_timeout_none(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(wr, selectors.EVENT_WRITE)
+
+        with self.assertTakesTime(upper=SHORT_SELECT):
+            self.assertEqual(1, len(s.select(timeout=None)))
+            self.assertEqual(1, len(s.select(timeout=None)))
+            self.assertEqual(1, len(s.select(timeout=None)))
+
+    def test_select_timeout_ready(self):
+        s, rd, wr = self.standard_setup()
+
+        with self.assertTakesTime(upper=SHORT_SELECT):
+            self.assertEqual(1, len(s.select(timeout=0)))
+            self.assertEqual(1, len(s.select(timeout=-1)))
+            self.assertEqual(1, len(s.select(timeout=0.001)))
+
+    def test_select_timeout_not_ready(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+
+        with self.assertTakesTime(upper=SHORT_SELECT):
+            self.assertEqual(0, len(s.select(timeout=0)))
+
+        with self.assertTakesTime(lower=SHORT_SELECT):
+            self.assertEqual(0, len(s.select(timeout=SHORT_SELECT)))
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_timing(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        key = s.register(rd, selectors.EVENT_READ)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, lambda *args: wr.send(b'x'))
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(upper=LONG_SELECT):
+            ready = s.select(LONG_SELECT)
+            self.assertEqual([(key, selectors.EVENT_READ)], ready)
```

Comment:
As a basic note about cleanliness, you should probably remove the actual assertion from the context manager in this kind of case.